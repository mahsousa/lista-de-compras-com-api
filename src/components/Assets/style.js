import styled from "styled-components";

export const Container = styled.div`
  display:flex;
  justify-content: center;
  align-items: center;
  background-color: #fff;
  border-radius:5px;
  margin: 20px;
  padding: 20px;

  @media (max-width:320px){
    margin:0 2px;
  }

  @media (min-width: 481px) and (max-width: 991px){
    margin: 2px;
    padding: 10px;
    justify-content: center;
  }
`;