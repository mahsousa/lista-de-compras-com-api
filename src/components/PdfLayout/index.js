import React from 'react';
import { Font, Image, Page, Text, View, Document, StyleSheet } from '@react-pdf/renderer';


Font.register({
  family: 'Poppins',
  src: 'http://fonts.gstatic.com/s/poppins/v1/hlvAxH6aIdOjWlLzgm0jqg.ttf',
});

// Create styles
const styles = StyleSheet.create({
  body: {
    paddingTop: 35,
    paddingHorizontal: 35,
  },
  titlePage: {
    fontSize: 25,
    padding: 20,
    marginBottom: 40,
    textAlign: 'center',
    backgroundColor: '#E4E4E4',
    color: '#808080',
  },
  contentTitle: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 20,
    padding: 5,
    fontSize: 16,
  },
  section: {
    fontFamily: 'Poppins',
    padding: 5,
  },
  informacoes: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    fontSize: 14,
  },
  infoTitulo: {
    width: 150,
  }
});



// Create Document Component
export const PdfLayout = () => {


  function getdata() {
    return JSON.parse(localStorage.getItem('transactionsList'));
  }


  function getdata() {
    return JSON.parse(localStorage.getItem('transactions'));
  }

  function renderRow(row) {
    return (
      <Document style={styles.informacoes}>
        <View style={styles.infoTitulo}>
          <Text style={styles.line}>{row.descricao}</Text>
        </View>
        <View style={styles.infoTitulo}>
          <Text style={styles.line}>{row.saida ? '-' : '+'}{row.quantia.toString().replace('.',',')}</Text>
        </View>
        <View style={styles.infoTitulo}>
          <Text style={styles.line}>{row.saida ? 'Despesa' : 'Novo Orçamento'}</Text>
        </View>
      </Document>
    )
  }

  function renderRowTotais(totalOrcamento, totalGasto, saldoDisponivel) {
    return (
      <Document style={styles.informacoes}>
        <View style={styles.infoTitulo}>
          <Text style={styles.line}>R${totalOrcamento.toFixed(2).toString().replace('.',',')}</Text>
        </View>
        <View style={styles.infoTitulo}>
          <Text style={styles.line}>R${totalGasto.toFixed(2).toString().replace('.',',')}</Text>
        </View>
        <View style={styles.infoTitulo}>
          <Text style={styles.line}>R${saldoDisponivel.toFixed(2).toString().replace('.',',')}</Text>
        </View>
      </Document>
    )
  }

  function prepararSessaoTotais(data){
    let totalOrcamento = 0, totalGasto = 0, saldoDisponivel = 0;
    for(let i = 0; i < data.length; i++){
      if(data[i].saida){
        totalGasto += parseFloat(data[i].quantia);
      }
      else{
        totalOrcamento += parseFloat(data[i].quantia);
      }
    }
    saldoDisponivel = totalOrcamento - totalGasto;

    rowsTotais.push(renderRowTotais(totalOrcamento, totalGasto, saldoDisponivel));

  }

  let rows = [];
  let rowsTotais = [];
  let data = getdata();

  if (data != null) {
    prepararSessaoTotais(data);

    for (let i = 0; i < data.length; i++) {
      rows.push(renderRow(data[i]));
    }
  }

  let date = new Date();
  let dataFormatada = ((date.getDate() )) + "." + ((date.getMonth() + 1)) + "." + date.getFullYear(); 

  return (
    <Document>
      <Page size="A4" style={styles.body}>
        <Text style={styles.titlePage}>
          Relação de Gastos - {dataFormatada}
        </Text>

        <View style={styles.contentTitle}>
          <View style={styles.infoTitulo}>
            <Text style={styles.line}>Total Orçamento:</Text>
          </View>
          <View style={styles.infoTitulo}>
            <Text style={styles.line}>Total Gasto:</Text>
          </View>
          <View style={styles.infoTitulo}>
            <Text style={styles.line}>Saldo Disponivel:</Text>
          </View>
        </View>
        <View style={styles.section}>
          {rowsTotais}
        </View>



        <View style={styles.contentTitle}>
          <View style={styles.infoTitulo}>
            <Text style={styles.line}>Descrição:</Text>
          </View>
          <View style={styles.infoTitulo}>
            <Text style={styles.line}>Valor:</Text>
          </View>
          <View style={styles.infoTitulo}>
            <Text style={styles.line}>E | S: </Text>
          </View>
        </View>
        <View style={styles.section}>
          {rows}
        </View>
      </Page>
    </Document>
  )
};

export default PdfLayout;