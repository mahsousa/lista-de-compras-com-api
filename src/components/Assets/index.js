import React from "react";
import * as C from "./style";
import { PDFDownloadLink } from '@react-pdf/renderer';
import PdfLayout from "../PdfLayout"

import {
  AiFillPrinter,
} from "react-icons/ai";

import {
  FaTrashAlt,
} from "react-icons/fa";


function Assets() {

  function clearAllStorage() {
    localStorage.removeItem("transactions");
    window.location.reload(true);

  }

  return (
    <>
      <C.Container>
        <PDFDownloadLink document={<PdfLayout />} fileName="somename.pdf">
          {({ loading }) =>
            loading ? 'Loading document...' : ''
          }
          <AiFillPrinter style={{
            color: "#215581",
            height: '40px',
            width: '40px',
            padding:'5px'
          }} />
        </PDFDownloadLink >
        <FaTrashAlt style={{
          color: "#215581",
          height: '40px',
          width: '30px',
          padding:'5px'
        }} onClick={clearAllStorage} />
      </C.Container>
    </>
  )

}

export default Assets;