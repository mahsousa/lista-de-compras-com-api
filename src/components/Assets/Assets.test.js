import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { PDFDownloadLink } from '@react-pdf/renderer';
import PdfLayout from "../PdfLayout"
import Assets from '.';


import {
  AiFillPrinter,
} from "react-icons/ai";

import {
  FaTrashAlt,
} from "react-icons/fa";


jest.mock("@react-pdf/renderer", () => {
  const React = require("react");

  const PDFDownloadLink = ({ document, fileName, children }) => {
    const mockOnClick = () => {
      // Fazer algo quando o link for clicado
    };

    return (
      <div onClick={mockOnClick}>
        {children}
      </div>
    );
  };

  const AiFillPrinter = () => {
    return <div>Mocked AiFillPrinter</div>;
  };

  return {
    __esModule: true,
    Font: {
      register: jest.fn(),
    },
    Image: jest.fn(),
    Page: jest.fn(({ children }) => <div>{children}</div>),
    Text: jest.fn(),
    View: jest.fn(),
    Document: jest.fn(({ children }) => <div>{children}</div>),
    StyleSheet: {
      create: jest.fn((styles) => styles),
    },
    PDFDownloadLink: PDFDownloadLink,
    AiFillPrinter: AiFillPrinter,
  };
});


it('Renderizar o componente corretamente', () => {
  const component = renderer.create(<Assets />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


it('Se a função clearAllStorage está sendo chamada quando FaTrashAlt é clicado', () => {
  const clearAllStorage = jest.fn();

  const { container } = render(
    <FaTrashAlt onClick={clearAllStorage} />
  );

  const faTrashAltElement = container.querySelector('svg');

  fireEvent.click(faTrashAltElement);

  expect(clearAllStorage).toHaveBeenCalledTimes(1);
});

it('chama o componente PdfLayout ao clicar no ícone de printer', () => {
  let pdfLayoutCalled = false;

  const MockPdfLayout = () => {
    pdfLayoutCalled = true;
    return null;
  };

  render(
      <PDFDownloadLink document={<MockPdfLayout />} className="print-btn" fileName="somename.pdf">
        {({ loading }) => (loading ? 'Loading document...' : '')}
        <AiFillPrinter
          data-testid="printer-icon"
          style={{
            color: '#215581',
            height: '40px',
            width: '40px',
            padding: '5px',
          }}
        />
      </PDFDownloadLink>
  );

  const printerIcon = screen.getByTestId('printer-icon');
  fireEvent.click(printerIcon);

  // Aguarda o próximo ciclo de microtarefa para verificar a atualização da variável
  setTimeout(() => {
    expect(pdfLayoutCalled).toBe(true);
  }, 0);
});


