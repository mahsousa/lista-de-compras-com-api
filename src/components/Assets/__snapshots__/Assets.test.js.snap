// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderizar o componente corretamente 1`] = `
<div
  className="sc-beqWaB bVwooW"
>
  <div
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": "#215581",
          "height": "40px",
          "padding": "5px",
          "width": "40px",
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M732 120c0-4.4-3.6-8-8-8H300c-4.4 0-8 3.6-8 8v148h440V120zm120 212H172c-44.2 0-80 35.8-80 80v328c0 17.7 14.3 32 32 32h168v132c0 4.4 3.6 8 8 8h424c4.4 0 8-3.6 8-8V772h168c17.7 0 32-14.3 32-32V412c0-44.2-35.8-80-80-80zM664 844H360V568h304v276zm164-360c0 4.4-3.6 8-8 8h-40c-4.4 0-8-3.6-8-8v-40c0-4.4 3.6-8 8-8h40c4.4 0 8 3.6 8 8v40z"
      />
    </svg>
  </div>
  <svg
    fill="currentColor"
    height="1em"
    onClick={[Function]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": "#215581",
        "height": "40px",
        "padding": "5px",
        "width": "30px",
      }
    }
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
    />
  </svg>
</div>
`;
